<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Bioinformatics Protocols - 2&nbsp; Bulk Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./functional.html" rel="next">
<link href="./pipelines.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./bulk.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Bulk Analysis</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Bioinformatics Protocols</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pipelines.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Glueing Together Pipelines</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./bulk.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Bulk Analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./functional.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Functional Analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./3D.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">3D Chromatin Organization</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./single-cell.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Single-Cell Analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Books and Resources</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#rna-seq" id="toc-rna-seq" class="nav-link active" data-scroll-target="#rna-seq"><span class="header-section-number">2.1</span> RNA-Seq</a>
  <ul class="collapse">
  <li><a href="#differential-gene-expression" id="toc-differential-gene-expression" class="nav-link" data-scroll-target="#differential-gene-expression"><span class="header-section-number">2.1.1</span> Differential Gene Expression</a></li>
  <li><a href="#deciding-groups-and-plotting" id="toc-deciding-groups-and-plotting" class="nav-link" data-scroll-target="#deciding-groups-and-plotting"><span class="header-section-number">2.1.2</span> Deciding Groups and Plotting</a></li>
  </ul></li>
  <li><a href="#chip-seq-and-atac-seq" id="toc-chip-seq-and-atac-seq" class="nav-link" data-scroll-target="#chip-seq-and-atac-seq"><span class="header-section-number">2.2</span> ChIP-Seq and ATAC-Seq</a>
  <ul class="collapse">
  <li><a href="#spike-in-normalization" id="toc-spike-in-normalization" class="nav-link" data-scroll-target="#spike-in-normalization"><span class="header-section-number">2.2.1</span> Spike-in normalization</a></li>
  <li><a href="#peak-calling" id="toc-peak-calling" class="nav-link" data-scroll-target="#peak-calling"><span class="header-section-number">2.2.2</span> Peak calling</a></li>
  <li><a href="#overlapping-peaks" id="toc-overlapping-peaks" class="nav-link" data-scroll-target="#overlapping-peaks"><span class="header-section-number">2.2.3</span> Overlapping peaks</a></li>
  <li><a href="#motif-enrichment" id="toc-motif-enrichment" class="nav-link" data-scroll-target="#motif-enrichment"><span class="header-section-number">2.2.4</span> Motif Enrichment</a></li>
  <li><a href="#plotting-heatmaps" id="toc-plotting-heatmaps" class="nav-link" data-scroll-target="#plotting-heatmaps"><span class="header-section-number">2.2.5</span> Plotting Heatmaps</a></li>
  <li><a href="#genomic-annotation" id="toc-genomic-annotation" class="nav-link" data-scroll-target="#genomic-annotation"><span class="header-section-number">2.2.6</span> Genomic Annotation</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-Bulk_Analysis" class="quarto-section-identifier"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Bulk Analysis</span></span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>This chapter is for the downstream analysis of the most common bulk NGS experiments we perform. For functional analysis that are common to all these methods, see next chapter.</p>
<section id="rna-seq" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="rna-seq"><span class="header-section-number">2.1</span> RNA-Seq</h2>
<p>While RPKM has some issues and is not the most correct quantification, for basic comparison, this is good enough to share with biologists. The nf-core pipeline will provide this value by default.</p>
<section id="differential-gene-expression" class="level3" data-number="2.1.1">
<h3 data-number="2.1.1" class="anchored" data-anchor-id="differential-gene-expression"><span class="header-section-number">2.1.1</span> Differential Gene Expression</h3>
<p>One of the most basic and common analysis on RNA-Seq. We use DESeq2 and their well documented <a href="https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">vignette</a> is worth reading from start to end for the beginner.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>While our pipeline originally used the <code>STAR --quantMode</code> to quantify genes, with switching to nf-core, we are also switching to <a href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf">STAR</a> alignment followed by <a href="https://github.com/deweylab/RSEM">RSEM</a> quantification.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> <span class="st">"./working_dir"</span>, <span class="at">pattern =</span> <span class="st">"*genes.results$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>rsem_results <span class="ot">&lt;-</span> <span class="fu">lapply</span>(files, read_delim)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>expected_counts_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(rsem_results, <span class="cf">function</span>(x) { x<span class="sc">$</span>expected_count })</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>expected_counts <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind, expected_counts_list) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>RSEM produces non-integer counts, and we can by-pass that by using <code>round()</code>. Alternatively, you can use <a href="https://bioconductor.org/packages/devel/bioc/vignettes/tximport/inst/doc/tximport.html">tximport</a> to read the files in.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>expected_counts <span class="ot">&lt;-</span> <span class="fu">round</span>(expected_counts) </span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(expected_counts) <span class="ot">&lt;-</span> rsem_results[[<span class="dv">1</span>]]<span class="sc">$</span>gene_id</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(expected_counts) <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_extract</span>(files, <span class="st">"mr</span><span class="sc">\\</span><span class="st">d+"</span>) <span class="co"># our RNA-seq samples usually start with MR</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>sampleTable <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">condition =</span> <span class="fu">factor</span>(<span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"control"</span>, <span class="st">"knockdown"</span>), <span class="at">each =</span> <span class="dv">3</span>)),</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>                          <span class="at">replicate =</span> <span class="fu">factor</span>(<span class="fu">rep</span>(<span class="fu">seq</span>(<span class="dv">1</span>,<span class="dv">3</span>))))</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(sampleTable) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(expected_counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>DESeq2</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(expected_counts, sampleTable, <span class="at">design =</span> <span class="sc">~</span>condition)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">counts</span>(dds)) <span class="sc">&gt;</span> <span class="dv">10</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> dds[keep,]</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">alpha =</span> <span class="fl">0.01</span>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Difference between rlog, vst and lfcShrink <a href="https://support.bioconductor.org/p/104615/" class="uri">https://support.bioconductor.org/p/104615/</a>.</p>
<p><strong>Plotting PCs</strong></p>
<p>DESeq2’s <code>plotPCA()</code> function will plot the top 500 most variable genes. The chunk below will plot all genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>degenes <span class="ot">&lt;-</span> res <span class="sc">%&gt;%</span> <span class="fu">subset</span>(padj <span class="sc">&lt;</span> <span class="fl">0.01</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>dds_rlog <span class="ot">&lt;-</span> <span class="fu">rlog</span>(dds)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>pca_data <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">assay</span>(dds_rlog)) <span class="sc">%&gt;%</span> <span class="fu">prcomp</span>()</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(pca_data, <span class="at">data =</span> sampleTable, <span class="at">colour =</span> <span class="st">"condition"</span>) <span class="sc">+</span> </span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="at">label =</span> <span class="fu">rownames</span>(sampleTable))</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>rlog_de <span class="ot">&lt;-</span> <span class="fu">assay</span>(dds_rlog) <span class="sc">%&gt;%</span> <span class="fu">subset</span>(<span class="fu">rownames</span>(dds_rlog) <span class="sc">%in%</span> <span class="fu">rownames</span>(degenes))</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>rlog_de_scaled <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(rlog_de)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After getting DEGs, you’d want to group the genes into biological functions. See <a href="functional.html#sec-enrichment"><span>Section&nbsp;3.2</span></a> for Over representation analysis (ORA) with GO and KEGG terms as well as Gene Set Enrichment Analysis (GSEA) with ranked genes.</p>
</section>
<section id="deciding-groups-and-plotting" class="level3" data-number="2.1.2">
<h3 data-number="2.1.2" class="anchored" data-anchor-id="deciding-groups-and-plotting"><span class="header-section-number">2.1.2</span> Deciding Groups and Plotting</h3>
<p>While you can use k-means manually to get seperate groups, <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/index.html">ComplexHeatmap</a> allows you to do so with more <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#heatmap-split">flexiblity</a> and get visualizations as well.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>mat_colors <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">replicate =</span> <span class="fu">c</span>(<span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Accent"</span>)),</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">condition =</span> <span class="fu">c</span>(<span class="fu">brewer.pal</span>(<span class="dv">6</span>, <span class="st">"Set1"</span>)))</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(mat_colors<span class="sc">$</span>replicate) <span class="ot">&lt;-</span> <span class="fu">unique</span>(sampleTable<span class="sc">$</span>replicate)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(mat_colors<span class="sc">$</span>condition) <span class="ot">&lt;-</span> sampleTable<span class="sc">$</span>condition</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>col_anno <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(<span class="at">df =</span> sampleTable,</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>                              <span class="at">which =</span> <span class="st">'col'</span>,</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>                              <span class="at">col =</span> mat_colors</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>hmap <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(rlog_de_scaled,</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>                   <span class="at">name =</span> <span class="st">"scaled"</span>,</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>                   </span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>                   <span class="co"># Row Params</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>                   <span class="at">show_row_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>                   <span class="at">row_title_rot=</span><span class="dv">0</span>,</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>                   <span class="at">cluster_row_slices =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>                   <span class="at">border =</span> <span class="cn">TRUE</span>,</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>                   <span class="at">row_km =</span> <span class="dv">2</span>, <span class="co"># split rows into 2 groups</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>                   </span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>                   <span class="co"># Column Params</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>                   <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>                   <span class="at">column_title =</span> <span class="st">"Rlog Transformed Expression for all DE genes"</span>,</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>                   <span class="at">top_annotation =</span> col_anno)</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>hmap <span class="ot">&lt;-</span> <span class="fu">draw</span>(hmap) <span class="co"># assigning so that k-means is only called once</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="fu">row_order</span>(hmap) <span class="co"># grab the different groups in rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>See <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#get-orders-and-dendrograms-from-heatmap">here</a> for why we assign <code>draw()</code>.</p>
</section>
</section>
<section id="chip-seq-and-atac-seq" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="chip-seq-and-atac-seq"><span class="header-section-number">2.2</span> ChIP-Seq and ATAC-Seq</h2>
<section id="spike-in-normalization" class="level3" data-number="2.2.1">
<h3 data-number="2.2.1" class="anchored" data-anchor-id="spike-in-normalization"><span class="header-section-number">2.2.1</span> Spike-in normalization</h3>
<p><a href="https://www.activemotif.com/documents/1977.pdf">Guidelines</a> on spike-in normalization from ActiveMotif.</p>
<ol type="1">
<li>Perform ChIP combining the Spike-in Chromatin, Spike-in Antibody, test chromatin and test antibody into the same tube for immunoprecipitation. We suggest using the guidelines provided for chromatin and antibody quantities based on the antibody target.</li>
<li>Follow ChIP with Next-Generation Sequencing.</li>
<li>Map ChIP-seq data to the test reference genome (e.g.&nbsp;human, mouse or other).</li>
<li>Map ChIP-seq data to the Drosophila reference genome.</li>
<li>Count uniquely aligning Drosophila sequence tags and identify the sample containing the least number of tags.</li>
<li>Compare Drosophila tag counts from other samples to the sample containing the least tags and generate a normalization factor for each comparison. (Sample 1 with lowest tag count / Sample 2) = Normalization factor</li>
<li>Downsample the tag counts of data sets proportional to the normalization factor determined</li>
</ol>
</section>
<section id="peak-calling" class="level3" data-number="2.2.2">
<h3 data-number="2.2.2" class="anchored" data-anchor-id="peak-calling"><span class="header-section-number">2.2.2</span> Peak calling</h3>
<p>Depending on if you use the nfcore’s pipeline or your own, you will have to call peaks. I use MACS2 and here are some details I’ve gathered.<br>
<a href="https://github.com/crazyhottommy/ChIP-seq-analysis/blob/master/part1.3_MACS2_peak_calling_details.md" class="uri">https://github.com/crazyhottommy/ChIP-seq-analysis/blob/master/part1.3_MACS2_peak_calling_details.md</a></p>
<p>Why I skip model building <a href="https://github.com/macs3-project/MACS/issues/391" class="uri">https://github.com/macs3-project/MACS/issues/391</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">macs2</span> callpeak <span class="at">-t</span> mxx_sorted.bam <span class="at">--outdir</span> macs/mxx <span class="at">-n</span> mxx <span class="at">-g</span> hs <span class="at">-q</span> 0.01 <span class="at">--nomodel</span> <span class="at">--shift</span> 0 <span class="at">--extsize</span> 250</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="overlapping-peaks" class="level3" data-number="2.2.3">
<h3 data-number="2.2.3" class="anchored" data-anchor-id="overlapping-peaks"><span class="header-section-number">2.2.3</span> Overlapping peaks</h3>
<p>After MACS2 peak-calling, we may want to see how many peaks overlap in different conditions. Even though it’s named <a href="http://homer.ucsd.edu/homer/ngs/mergePeaks.html">mergePeaks</a>, you will be able to get overlapping statistics from this. Be mindful of long path names as mergePeaks will produce errors.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mergePeaks</span> <span class="at">-d</span> 100 pu1.peaks cebp.peaks <span class="at">-prefix</span> mmm <span class="at">-venn</span> venn.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>The -d flag changes the unique peaks to 100 bp each and keep the shared peaks same size.<br>
To get literal 1bp overlap, just omit -d argument altogether</p>
</div>
</div>
<p>The easiest way to plot this result is with <a href="https://github.com/js229/Vennerable">Vennerable</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> tot[<span class="dv">1</span>]</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> tot[<span class="dv">2</span>]</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>ab <span class="ot">&lt;-</span> tot[<span class="dv">3</span>]</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>venn_obj <span class="ot">&lt;-</span> <span class="fu">createVennObj</span>(<span class="at">nSets =</span> <span class="dv">2</span>, <span class="at">sNames =</span> <span class="fu">c</span>(<span class="st">"m1043"</span>, <span class="st">"m1044"</span>), <span class="co"># names in order of first to last</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                     <span class="at">sSizes =</span> <span class="fu">c</span>(<span class="dv">0</span>, a, b ,ab))</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>vp <span class="ot">&lt;-</span> <span class="fu">plotVenn</span>(<span class="at">nVennObj =</span> venn_obj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This gets tedious with more than 2 sets. See code below. You should not make Venn Diagrams with more than 4 sets. An upset plot is better in that scenario.</p>
<div class="cell">
<details>
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 3 sets -----</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> tot[<span class="dv">1</span>]</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> tot[<span class="dv">2</span>]</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>ab <span class="ot">&lt;-</span> tot[<span class="dv">3</span>]</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>c <span class="ot">&lt;-</span> tot[<span class="dv">4</span>]</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>ac <span class="ot">&lt;-</span> tot[<span class="dv">5</span>]</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>bc <span class="ot">&lt;-</span> tot[<span class="dv">6</span>]</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>abc <span class="ot">&lt;-</span> tot[<span class="dv">7</span>]</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co"># 4 sets -------</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> tot[<span class="dv">1</span>]</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> tot[<span class="dv">2</span>]</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>ab <span class="ot">&lt;-</span> tot[<span class="dv">3</span>]Ve</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>c <span class="ot">&lt;-</span> tot[<span class="dv">4</span>]</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>ac <span class="ot">&lt;-</span> tot[<span class="dv">5</span>]</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>bc <span class="ot">&lt;-</span> tot[<span class="dv">6</span>]</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>abc <span class="ot">&lt;-</span> tot[<span class="dv">7</span>]</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> tot[<span class="dv">8</span>]</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>ad <span class="ot">&lt;-</span> tot[<span class="dv">9</span>]</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>bd <span class="ot">&lt;-</span> tot[<span class="dv">10</span>]</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>abd <span class="ot">&lt;-</span> tot[<span class="dv">11</span>]</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>cd <span class="ot">&lt;-</span> tot[<span class="dv">12</span>]</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a>acd <span class="ot">&lt;-</span> tot[<span class="dv">13</span>]</span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>bcd <span class="ot">&lt;-</span> tot[<span class="dv">14</span>]</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>abcd <span class="ot">&lt;-</span> tot[<span class="dv">15</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>This is a more automated way of reading in the data, here using 4 sets as an example.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>venn <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"results/output_nepc.bedpe"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>venn<span class="sc">$</span>alpha <span class="ot">&lt;-</span> <span class="fu">apply</span>(venn, <span class="dv">1</span>, <span class="cf">function</span>(x) {</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  sets <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  selected_sets <span class="ot">&lt;-</span> sets[<span class="fu">which</span>(x <span class="sc">==</span> <span class="st">"X"</span>)]</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste</span>(selected_sets, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>order_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"0"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"AB"</span>, <span class="st">"C"</span>, <span class="st">"AC"</span>, <span class="st">"BC"</span>, <span class="st">"ABC"</span>, <span class="st">"D"</span>, <span class="st">"AD"</span>, <span class="st">"BD"</span>, <span class="st">"ABD"</span>, <span class="st">"CD"</span>, <span class="st">"ACD"</span>, <span class="st">"BCD"</span>, <span class="st">"ABCD"</span>)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>venn<span class="sc">$</span>alpha <span class="ot">&lt;-</span> <span class="fu">factor</span>(venn<span class="sc">$</span>alpha, <span class="at">levels =</span> order_vector)</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Sort the data frame based on the factor levels</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>venn <span class="ot">&lt;-</span> venn[<span class="fu">order</span>(venn<span class="sc">$</span>alpha), ]</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>venn_plot <span class="ot">&lt;-</span> <span class="fu">Venn</span>(<span class="at">SetNames =</span> <span class="fu">c</span>(<span class="st">"93"</span>, <span class="st">"145.1"</span>, <span class="st">"145.2"</span>, <span class="st">"nci"</span>), </span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>                  <span class="at">Weight =</span> <span class="fu">c</span>(<span class="dv">0</span>, venn<span class="sc">$</span>Features))</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(venn_plot, <span class="at">type =</span> <span class="st">"ellipses"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Creating upset plots <a href="https://github.com/hms-dbmi/UpSetR" class="uri">https://github.com/hms-dbmi/UpSetR</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(UpSetR)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>venn<span class="sc">$</span>Sets <span class="ot">&lt;-</span> <span class="fu">apply</span>(venn[, <span class="sc">-</span><span class="dv">1</span>], <span class="dv">1</span>, <span class="cf">function</span>(x) {</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  sets <span class="ot">&lt;-</span> <span class="fu">colnames</span>(venn)[<span class="dv">2</span><span class="sc">:</span><span class="dv">17</span>]</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  selected_sets <span class="ot">&lt;-</span> sets[<span class="fu">which</span>(x <span class="sc">==</span> <span class="st">"X"</span>)]</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste</span>(selected_sets, <span class="at">collapse =</span> <span class="st">"&amp;"</span>)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>upset_input <span class="ot">&lt;-</span> <span class="fu">c</span>(venn<span class="sc">$</span>Features)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(upset_input) <span class="ot">&lt;-</span> venn<span class="sc">$</span>Sets</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="fu">upset</span>(<span class="fu">fromExpression</span>(upset_input), <span class="at">nsets =</span> <span class="dv">16</span>, <span class="at">nintersects =</span> <span class="dv">100</span>, <span class="at">number.angles =</span> <span class="dv">45</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A sample script to automate this process. I haven’t incorporated the above part into this script.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>mergePeaksandplot.sh</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load homer/4.10</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load R/4.0.3</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="ex">mergePeaks</span> control_peaks.narrowPeak purturbed_peaks.narrowPeak <span class="at">-prefix</span> ov <span class="at">-venn</span> venn.txt</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># n_way script, venn.txt, last sample label, middle samples label, first sample label, txt file name (no extension)</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">--vanilla</span> venn2way.R venn.txt purturbed control venn</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>The accompanying R file:</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>venn2way.R</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Vennerable)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>args <span class="ot">&lt;-</span> <span class="fu">commandArgs</span>(<span class="at">trailingOnly =</span> <span class="cn">TRUE</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>venn <span class="ot">&lt;-</span> <span class="fu">read.table</span>(args[<span class="dv">1</span>], <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>tot <span class="ot">&lt;-</span> venn[[Total]]</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> tot[<span class="dv">1</span>]</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> tot[<span class="dv">2</span>]</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>ab <span class="ot">&lt;-</span> tot[<span class="dv">3</span>]</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>venner <span class="ot">&lt;-</span> <span class="fu">Venn</span>(<span class="at">SetNames =</span> <span class="fu">c</span>(args[<span class="dv">3</span>], args[<span class="dv">2</span>]), <span class="co"># opposite labelling to nVennR, going from last sample to first</span></span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>               <span class="at">Weight =</span> <span class="fu">c</span>(<span class="dv">0</span>, b, a, ab))</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">paste0</span>(args[<span class="dv">4</span>], <span class="st">".png"</span>))</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(venner)</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>To make Venn Diagrams that are more accurately weighted, use <a href="https://cran.r-project.org/web/packages/nVennR/index.html">nVennR</a>. Sadly, it looks like it’s been removed from CRAN since the last time I’ve installed it.</p>
<p>If you have replicates, overlapping peaks can be obtained via packages like <a href="https://cran.r-project.org/web/packages/MAnorm2/index.html">MAnorm2</a>. Read more about dealing with replicates <a href="https://ro-che.info/articles/2018-07-11-chip-seq-consensus">here</a>.</p>
</section>
<section id="motif-enrichment" class="level3" data-number="2.2.4">
<h3 data-number="2.2.4" class="anchored" data-anchor-id="motif-enrichment"><span class="header-section-number">2.2.4</span> Motif Enrichment</h3>
<p>Homer <code>mergePeaks</code> is able to take the narrowPeaks format as input, but if you’re doing homer <code>mergePeaks</code> first and then <code>findMotifs</code>, you need to change the Homer <code>mergePeaks</code> output to exclude the header line first. This also applies to deepTools heatmaps.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">findMotifsGenome.pl</span> peak_or_bed /projects/p20023/Viriya/software/Homer4.10/data/genomes/hg38/ output_dir <span class="at">-size</span> 200 </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>-size 200 is the default, and is calculated from the center of the peaks. If your peaks are bigger and you wish to use the entire region, use <code>-size given</code>. However, when the regions are too large, motifs will not be significantly enriched.</p>
</section>
<section id="plotting-heatmaps" class="level3" data-number="2.2.5">
<h3 data-number="2.2.5" class="anchored" data-anchor-id="plotting-heatmaps"><span class="header-section-number">2.2.5</span> Plotting Heatmaps</h3>
<p>Here I’m showing a sample <a href="https://deeptools.readthedocs.io/en/latest/">deepTools</a> script to plot heatmaps. First you need bigwig files. <code>bamCoverage</code> offers normalization by scaling factor, Reads Per Kilobase per Million mapped reads (RPKM), counts per million (CPM), bins per million mapped reads (BPM) and 1x depth (reads per genome coverage, RPGC).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">bamCoverage</span> <span class="at">--bam</span> m<span class="va">${i}</span>_sorted.bam <span class="at">--outFileName</span> m<span class="va">${i}</span>.bw <span class="at">--normalizeUsing</span> CPM <span class="dt">\</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>--extendReads 200 <span class="at">--numberOfProcessors</span> 6 <span class="at">--binSize</span> 20</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>I find having paths defined at the top of the script makes it less likely to have mistakes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="va">day3</span><span class="op">=</span><span class="st">"m1015_1055_hg38/foxa2/CPM"</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="va">nci</span><span class="op">=</span><span class="st">"m1015_1055_hg38/CPM"</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="va">chip</span><span class="op">=</span><span class="st">"../Viriya/analysis/foxa2/chip-seq"</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="ex">computeMatrix</span> reference-point <span class="at">--referencePoint</span> center <span class="at">-S</span> <span class="dt">\</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="va">$day3</span>/m674.bw <span class="va">$day3</span>/m676.bw <span class="va">$tf</span>/m914.bw <span class="va">$tf</span>/m921.bw <span class="va">$nci</span>/m1043.bw <span class="va">$nci</span>/m1044.bw <span class="dt">\</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>-R <span class="va">$chip</span>/clustering/output/1.bed <span class="dt">\</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a><span class="va">$chip</span>/clustering/output/2.bed <span class="dt">\</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="va">$chip</span>/clustering/output/3.bed <span class="dt">\</span></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a><span class="va">$chip</span>/clustering/output/4.bed <span class="dt">\</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>-a 3000 <span class="at">-b</span> 3000 <span class="at">-o</span> <span class="va">$chip</span>/heatmaps/scaled_FOXA2_sites.npz <span class="dt">\</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>--samplesLabel m674 m676 m914 m921 m1043 m1044 <span class="dt">\</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>-p max <span class="at">--blackListFileName</span> <span class="va">$chip</span>/ENCFF356LFX.bed.gz</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a><span class="ex">plotHeatmap</span> <span class="at">-m</span> <span class="va">$chip</span>/FOXA2_6_samples/heatmaps/scaled_FOXA2_sites.npz <span class="dt">\</span></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>-o <span class="va">$chip</span>/FOXA2_6_samples/heatmaps/scaled_FOXA2_sites.pdf <span class="dt">\</span></span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>--colorMap Blues</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>--sortUsingSamples</code> is not 0 indexed. 1st sample is 1. <code>--sortUsingSamples</code> is also usable in plotHeatmap so we can save that to there for more flexible plotting.</p>
<p><code>--regionsLabel</code>: “xxx binding sites n=2390” put it in quotes if you don’t want to key in escape too many times</p>
<p><code>--sampleLabels</code>: if you’re sure you’re not going to change any labels, put it in makeheatmap step because fewer things to type when redoing the heatmap multiple times due to scaling etc, therefore fewer mistakes. otherwise put it in plotHeatmap.</p>
<p>In legends: “Pol II” spaces and parantheses need to be escaped, but not + signs.</p>
<p>Use BED6 format annd not the BED3 if you care about strandedness <a href="https://github.com/deeptools/deepTools/issues/886" class="uri">https://github.com/deeptools/deepTools/issues/886</a>.</p>
<p>To get sorted output bed files, use the <code>--sortedOutRegions</code> at the plotHeatmap step, not computeMatrix.<br>
Use <code>--clusterUsingSamples</code> for a more robust delineation.</p>
</section>
<section id="genomic-annotation" class="level3" data-number="2.2.6">
<h3 data-number="2.2.6" class="anchored" data-anchor-id="genomic-annotation"><span class="header-section-number">2.2.6</span> Genomic Annotation</h3>
<p><a href="https://bioconductor.org/packages/release/bioc/html/ChIPseeker.html">ChIPSeeker</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TxDb.Hsapiens.UCSC.hg38.knownGene)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>promoter <span class="ot">&lt;-</span> <span class="fu">getPromoters</span>(<span class="at">TxDb=</span>txdb, <span class="at">upstream=</span><span class="dv">3000</span>, <span class="at">downstream=</span><span class="dv">3000</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>mxxx <span class="ot">&lt;-</span> <span class="fu">readPeakFile</span>(<span class="st">"/path_to_narrowPeak/mxxx_peaks.narrowPeak"</span>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>tagMatrix <span class="ot">&lt;-</span> <span class="fu">getTagMatrix</span>(mxxx, <span class="at">windows =</span> promoter) <span class="co"># usually needs an interactive job</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>peakAnno <span class="ot">&lt;-</span> <span class="fu">annotatePeak</span>(mxxx, <span class="at">tssRegion =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">3000</span>,<span class="dv">3000</span>), <span class="at">TxDb =</span> txdb, <span class="at">annoDb =</span> <span class="st">"org.Hs.eg.db"</span>)</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAnnoPie</span>(peakAnno, <span class="at">main =</span> <span class="st">"mxxx"</span>, <span class="at">line =</span> <span class="sc">-</span><span class="dv">6</span>)</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="fu">vennpie</span>(peakAnno, <span class="at">r =</span> <span class="fl">0.1</span>)</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="fu">upsetplot</span>(peakAnno) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"mxxx"</span>)</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAnnoBar</span>(peakAnno)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>annotatePeak</code> will default to chosing one gene per region. This is fine for small binding sites but for larger regions use <code>seq2gene</code> as it will map genomic regions in a many-to-many manner.</p>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
      }
    }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./pipelines.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Glueing Together Pipelines</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./functional.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Functional Analysis</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb18" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="an">execute:</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="co">  eval: false</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="co">    link-external-newwindow: true</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="fu"># Bulk Analysis {#sec-Bulk_Analysis}</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>This chapter is for the downstream analysis of the most common bulk NGS experiments we perform. For functional analysis that are common to all these methods, see next chapter.</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="fu">## RNA-Seq</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>While RPKM has some issues and is not the most correct quantification, for basic comparison, this is good enough to share with biologists. The nf-core pipeline will provide this value by default.</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a><span class="fu">### Differential Gene Expression</span></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>One of the most basic and common analysis on RNA-Seq. We use DESeq2 and their well documented <span class="co">[</span><span class="ot">vignette</span><span class="co">](https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)</span> is worth reading from start to end for the beginner.</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a>While our pipeline originally used the <span class="in">`STAR --quantMode`</span> to quantify genes, with switching to nf-core, we are also switching to <span class="co">[</span><span class="ot">STAR</span><span class="co">](https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf)</span> alignment followed by <span class="co">[</span><span class="ot">RSEM</span><span class="co">](https://github.com/deweylab/RSEM)</span> quantification.</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> <span class="st">"./working_dir"</span>, <span class="at">pattern =</span> <span class="st">"*genes.results$"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a>rsem_results <span class="ot">&lt;-</span> <span class="fu">lapply</span>(files, read_delim)</span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a>expected_counts_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(rsem_results, <span class="cf">function</span>(x) { x<span class="sc">$</span>expected_count })</span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a>expected_counts <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind, expected_counts_list) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-36"><a href="#cb18-36" aria-hidden="true" tabindex="-1"></a>RSEM produces non-integer counts, and we can by-pass that by using <span class="in">`round()`</span>. Alternatively, you can use <span class="co">[</span><span class="ot">tximport</span><span class="co">](https://bioconductor.org/packages/devel/bioc/vignettes/tximport/inst/doc/tximport.html)</span> to read the files in.</span>
<span id="cb18-37"><a href="#cb18-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-40"><a href="#cb18-40" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb18-41"><a href="#cb18-41" aria-hidden="true" tabindex="-1"></a>expected_counts <span class="ot">&lt;-</span> <span class="fu">round</span>(expected_counts) </span>
<span id="cb18-42"><a href="#cb18-42" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(expected_counts) <span class="ot">&lt;-</span> rsem_results[[<span class="dv">1</span>]]<span class="sc">$</span>gene_id</span>
<span id="cb18-43"><a href="#cb18-43" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(expected_counts) <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_extract</span>(files, <span class="st">"mr</span><span class="sc">\\</span><span class="st">d+"</span>) <span class="co"># our RNA-seq samples usually start with MR</span></span>
<span id="cb18-44"><a href="#cb18-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-45"><a href="#cb18-45" aria-hidden="true" tabindex="-1"></a>sampleTable <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">condition =</span> <span class="fu">factor</span>(<span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"control"</span>, <span class="st">"knockdown"</span>), <span class="at">each =</span> <span class="dv">3</span>)),</span>
<span id="cb18-46"><a href="#cb18-46" aria-hidden="true" tabindex="-1"></a>                          <span class="at">replicate =</span> <span class="fu">factor</span>(<span class="fu">rep</span>(<span class="fu">seq</span>(<span class="dv">1</span>,<span class="dv">3</span>))))</span>
<span id="cb18-47"><a href="#cb18-47" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(sampleTable) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(expected_counts)</span>
<span id="cb18-48"><a href="#cb18-48" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-49"><a href="#cb18-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-50"><a href="#cb18-50" aria-hidden="true" tabindex="-1"></a>**DESeq2**</span>
<span id="cb18-51"><a href="#cb18-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-54"><a href="#cb18-54" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb18-55"><a href="#cb18-55" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(expected_counts, sampleTable, <span class="at">design =</span> <span class="sc">~</span>condition)</span>
<span id="cb18-56"><a href="#cb18-56" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">counts</span>(dds)) <span class="sc">&gt;</span> <span class="dv">10</span></span>
<span id="cb18-57"><a href="#cb18-57" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> dds[keep,]</span>
<span id="cb18-58"><a href="#cb18-58" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span>
<span id="cb18-59"><a href="#cb18-59" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">alpha =</span> <span class="fl">0.01</span>)</span>
<span id="cb18-60"><a href="#cb18-60" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(res)</span>
<span id="cb18-61"><a href="#cb18-61" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-62"><a href="#cb18-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-63"><a href="#cb18-63" aria-hidden="true" tabindex="-1"></a>Difference between rlog, vst and lfcShrink <span class="ot">&lt;https://support.bioconductor.org/p/104615/&gt;</span>.</span>
<span id="cb18-64"><a href="#cb18-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-65"><a href="#cb18-65" aria-hidden="true" tabindex="-1"></a>**Plotting PCs**</span>
<span id="cb18-66"><a href="#cb18-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-67"><a href="#cb18-67" aria-hidden="true" tabindex="-1"></a>DESeq2's <span class="in">`plotPCA()`</span> function will plot the top 500 most variable genes. The chunk below will plot all genes.</span>
<span id="cb18-68"><a href="#cb18-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-71"><a href="#cb18-71" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb18-72"><a href="#cb18-72" aria-hidden="true" tabindex="-1"></a>degenes <span class="ot">&lt;-</span> res <span class="sc">%&gt;%</span> <span class="fu">subset</span>(padj <span class="sc">&lt;</span> <span class="fl">0.01</span>)</span>
<span id="cb18-73"><a href="#cb18-73" aria-hidden="true" tabindex="-1"></a>dds_rlog <span class="ot">&lt;-</span> <span class="fu">rlog</span>(dds)</span>
<span id="cb18-74"><a href="#cb18-74" aria-hidden="true" tabindex="-1"></a>pca_data <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">assay</span>(dds_rlog)) <span class="sc">%&gt;%</span> <span class="fu">prcomp</span>()</span>
<span id="cb18-75"><a href="#cb18-75" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(pca_data, <span class="at">data =</span> sampleTable, <span class="at">colour =</span> <span class="st">"condition"</span>) <span class="sc">+</span> </span>
<span id="cb18-76"><a href="#cb18-76" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="at">label =</span> <span class="fu">rownames</span>(sampleTable))</span>
<span id="cb18-77"><a href="#cb18-77" aria-hidden="true" tabindex="-1"></a>rlog_de <span class="ot">&lt;-</span> <span class="fu">assay</span>(dds_rlog) <span class="sc">%&gt;%</span> <span class="fu">subset</span>(<span class="fu">rownames</span>(dds_rlog) <span class="sc">%in%</span> <span class="fu">rownames</span>(degenes))</span>
<span id="cb18-78"><a href="#cb18-78" aria-hidden="true" tabindex="-1"></a>rlog_de_scaled <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">scale</span>(<span class="fu">t</span>(rlog_de)))</span>
<span id="cb18-79"><a href="#cb18-79" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-80"><a href="#cb18-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-81"><a href="#cb18-81" aria-hidden="true" tabindex="-1"></a>After getting DEGs, you'd want to group the genes into biological functions. See @sec-enrichment for Over representation analysis (ORA) with GO and KEGG terms as well as Gene Set Enrichment Analysis (GSEA) with ranked genes.</span>
<span id="cb18-82"><a href="#cb18-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-83"><a href="#cb18-83" aria-hidden="true" tabindex="-1"></a><span class="fu">### Deciding Groups and Plotting</span></span>
<span id="cb18-84"><a href="#cb18-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-85"><a href="#cb18-85" aria-hidden="true" tabindex="-1"></a>While you can use k-means manually to get seperate groups, <span class="co">[</span><span class="ot">ComplexHeatmap</span><span class="co">](https://jokergoo.github.io/ComplexHeatmap-reference/book/index.html)</span> allows you to do so with more <span class="co">[</span><span class="ot">flexiblity</span><span class="co">](https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#heatmap-split)</span> and get visualizations as well.</span>
<span id="cb18-86"><a href="#cb18-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-89"><a href="#cb18-89" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb18-90"><a href="#cb18-90" aria-hidden="true" tabindex="-1"></a>mat_colors <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb18-91"><a href="#cb18-91" aria-hidden="true" tabindex="-1"></a>  <span class="at">replicate =</span> <span class="fu">c</span>(<span class="fu">brewer.pal</span>(<span class="dv">3</span>, <span class="st">"Accent"</span>)),</span>
<span id="cb18-92"><a href="#cb18-92" aria-hidden="true" tabindex="-1"></a>  <span class="at">condition =</span> <span class="fu">c</span>(<span class="fu">brewer.pal</span>(<span class="dv">6</span>, <span class="st">"Set1"</span>)))</span>
<span id="cb18-93"><a href="#cb18-93" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(mat_colors<span class="sc">$</span>replicate) <span class="ot">&lt;-</span> <span class="fu">unique</span>(sampleTable<span class="sc">$</span>replicate)</span>
<span id="cb18-94"><a href="#cb18-94" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(mat_colors<span class="sc">$</span>condition) <span class="ot">&lt;-</span> sampleTable<span class="sc">$</span>condition</span>
<span id="cb18-95"><a href="#cb18-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-96"><a href="#cb18-96" aria-hidden="true" tabindex="-1"></a>col_anno <span class="ot">&lt;-</span> <span class="fu">HeatmapAnnotation</span>(<span class="at">df =</span> sampleTable,</span>
<span id="cb18-97"><a href="#cb18-97" aria-hidden="true" tabindex="-1"></a>                              <span class="at">which =</span> <span class="st">'col'</span>,</span>
<span id="cb18-98"><a href="#cb18-98" aria-hidden="true" tabindex="-1"></a>                              <span class="at">col =</span> mat_colors</span>
<span id="cb18-99"><a href="#cb18-99" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb18-100"><a href="#cb18-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-101"><a href="#cb18-101" aria-hidden="true" tabindex="-1"></a>hmap <span class="ot">&lt;-</span> <span class="fu">Heatmap</span>(rlog_de_scaled,</span>
<span id="cb18-102"><a href="#cb18-102" aria-hidden="true" tabindex="-1"></a>                   <span class="at">name =</span> <span class="st">"scaled"</span>,</span>
<span id="cb18-103"><a href="#cb18-103" aria-hidden="true" tabindex="-1"></a>                   </span>
<span id="cb18-104"><a href="#cb18-104" aria-hidden="true" tabindex="-1"></a>                   <span class="co"># Row Params</span></span>
<span id="cb18-105"><a href="#cb18-105" aria-hidden="true" tabindex="-1"></a>                   <span class="at">show_row_names =</span> <span class="cn">FALSE</span>,</span>
<span id="cb18-106"><a href="#cb18-106" aria-hidden="true" tabindex="-1"></a>                   <span class="at">row_title_rot=</span><span class="dv">0</span>,</span>
<span id="cb18-107"><a href="#cb18-107" aria-hidden="true" tabindex="-1"></a>                   <span class="at">cluster_row_slices =</span> <span class="cn">FALSE</span>,</span>
<span id="cb18-108"><a href="#cb18-108" aria-hidden="true" tabindex="-1"></a>                   <span class="at">border =</span> <span class="cn">TRUE</span>,</span>
<span id="cb18-109"><a href="#cb18-109" aria-hidden="true" tabindex="-1"></a>                   <span class="at">row_km =</span> <span class="dv">2</span>, <span class="co"># split rows into 2 groups</span></span>
<span id="cb18-110"><a href="#cb18-110" aria-hidden="true" tabindex="-1"></a>                   </span>
<span id="cb18-111"><a href="#cb18-111" aria-hidden="true" tabindex="-1"></a>                   <span class="co"># Column Params</span></span>
<span id="cb18-112"><a href="#cb18-112" aria-hidden="true" tabindex="-1"></a>                   <span class="at">cluster_columns =</span> <span class="cn">FALSE</span>,</span>
<span id="cb18-113"><a href="#cb18-113" aria-hidden="true" tabindex="-1"></a>                   <span class="at">column_title =</span> <span class="st">"Rlog Transformed Expression for all DE genes"</span>,</span>
<span id="cb18-114"><a href="#cb18-114" aria-hidden="true" tabindex="-1"></a>                   <span class="at">top_annotation =</span> col_anno)</span>
<span id="cb18-115"><a href="#cb18-115" aria-hidden="true" tabindex="-1"></a>hmap <span class="ot">&lt;-</span> <span class="fu">draw</span>(hmap) <span class="co"># assigning so that k-means is only called once</span></span>
<span id="cb18-116"><a href="#cb18-116" aria-hidden="true" tabindex="-1"></a><span class="fu">row_order</span>(hmap) <span class="co"># grab the different groups in rows</span></span>
<span id="cb18-117"><a href="#cb18-117" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-118"><a href="#cb18-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-119"><a href="#cb18-119" aria-hidden="true" tabindex="-1"></a>See <span class="co">[</span><span class="ot">here</span><span class="co">](https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html#get-orders-and-dendrograms-from-heatmap)</span> for why we assign <span class="in">`draw()`</span>.</span>
<span id="cb18-120"><a href="#cb18-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-121"><a href="#cb18-121" aria-hidden="true" tabindex="-1"></a><span class="fu">## ChIP-Seq and ATAC-Seq</span></span>
<span id="cb18-122"><a href="#cb18-122" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-123"><a href="#cb18-123" aria-hidden="true" tabindex="-1"></a><span class="fu">### Spike-in normalization</span></span>
<span id="cb18-124"><a href="#cb18-124" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Guidelines</span><span class="co">](https://www.activemotif.com/documents/1977.pdf)</span> on spike-in normalization from ActiveMotif.</span>
<span id="cb18-125"><a href="#cb18-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-126"><a href="#cb18-126" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Perform ChIP combining the Spike-in Chromatin, Spike-in Antibody, test chromatin and test antibody into the same tube for immunoprecipitation. We suggest using the guidelines provided for chromatin and antibody quantities based on the antibody target.</span>
<span id="cb18-127"><a href="#cb18-127" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Follow ChIP with Next-Generation Sequencing.</span>
<span id="cb18-128"><a href="#cb18-128" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Map ChIP-seq data to the test reference genome (e.g. human, mouse or other).</span>
<span id="cb18-129"><a href="#cb18-129" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Map ChIP-seq data to the Drosophila reference genome.</span>
<span id="cb18-130"><a href="#cb18-130" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>Count uniquely aligning Drosophila sequence tags and identify the sample containing the least number of tags.</span>
<span id="cb18-131"><a href="#cb18-131" aria-hidden="true" tabindex="-1"></a><span class="ss">6. </span>Compare Drosophila tag counts from other samples to the sample containing the least tags and generate a normalization factor for</span>
<span id="cb18-132"><a href="#cb18-132" aria-hidden="true" tabindex="-1"></a>each comparison.</span>
<span id="cb18-133"><a href="#cb18-133" aria-hidden="true" tabindex="-1"></a> (Sample 1 with lowest tag count / Sample 2) = Normalization factor</span>
<span id="cb18-134"><a href="#cb18-134" aria-hidden="true" tabindex="-1"></a><span class="ss">7. </span>Downsample the tag counts of data sets proportional to the normalization factor determined</span>
<span id="cb18-135"><a href="#cb18-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-136"><a href="#cb18-136" aria-hidden="true" tabindex="-1"></a><span class="fu">### Peak calling</span></span>
<span id="cb18-137"><a href="#cb18-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-138"><a href="#cb18-138" aria-hidden="true" tabindex="-1"></a>Depending on if you use the nfcore's pipeline or your own, you will have to call peaks. I use MACS2 and here are some details I've gathered.\</span>
<span id="cb18-139"><a href="#cb18-139" aria-hidden="true" tabindex="-1"></a><span class="ot">&lt;https://github.com/crazyhottommy/ChIP-seq-analysis/blob/master/part1.3_MACS2_peak_calling_details.md&gt;</span></span>
<span id="cb18-140"><a href="#cb18-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-141"><a href="#cb18-141" aria-hidden="true" tabindex="-1"></a>Why I skip model building <span class="ot">&lt;https://github.com/macs3-project/MACS/issues/391&gt;</span></span>
<span id="cb18-142"><a href="#cb18-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-145"><a href="#cb18-145" aria-hidden="true" tabindex="-1"></a><span class="in">```{bash}</span></span>
<span id="cb18-146"><a href="#cb18-146" aria-hidden="true" tabindex="-1"></a><span class="ex">macs2</span> callpeak <span class="at">-t</span> mxx_sorted.bam <span class="at">--outdir</span> macs/mxx <span class="at">-n</span> mxx <span class="at">-g</span> hs <span class="at">-q</span> 0.01 <span class="at">--nomodel</span> <span class="at">--shift</span> 0 <span class="at">--extsize</span> 250</span>
<span id="cb18-147"><a href="#cb18-147" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-148"><a href="#cb18-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-149"><a href="#cb18-149" aria-hidden="true" tabindex="-1"></a><span class="fu">### Overlapping peaks</span></span>
<span id="cb18-150"><a href="#cb18-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-151"><a href="#cb18-151" aria-hidden="true" tabindex="-1"></a>After MACS2 peak-calling, we may want to see how many peaks overlap in different conditions. Even though it's named <span class="co">[</span><span class="ot">mergePeaks</span><span class="co">](http://homer.ucsd.edu/homer/ngs/mergePeaks.html)</span>, you will be able to get overlapping statistics from this. Be mindful of long path names as mergePeaks will produce errors.</span>
<span id="cb18-152"><a href="#cb18-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-155"><a href="#cb18-155" aria-hidden="true" tabindex="-1"></a><span class="in">```{bash}</span></span>
<span id="cb18-156"><a href="#cb18-156" aria-hidden="true" tabindex="-1"></a><span class="ex">mergePeaks</span> <span class="at">-d</span> 100 pu1.peaks cebp.peaks <span class="at">-prefix</span> mmm <span class="at">-venn</span> venn.txt</span>
<span id="cb18-157"><a href="#cb18-157" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-158"><a href="#cb18-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-159"><a href="#cb18-159" aria-hidden="true" tabindex="-1"></a>::: callout-tip</span>
<span id="cb18-160"><a href="#cb18-160" aria-hidden="true" tabindex="-1"></a>The -d flag changes the unique peaks to 100 bp each and keep the shared peaks same size.\</span>
<span id="cb18-161"><a href="#cb18-161" aria-hidden="true" tabindex="-1"></a>To get literal 1bp overlap, just omit -d argument altogether</span>
<span id="cb18-162"><a href="#cb18-162" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb18-163"><a href="#cb18-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-164"><a href="#cb18-164" aria-hidden="true" tabindex="-1"></a>The easiest way to plot this result is with <span class="co">[</span><span class="ot">Vennerable</span><span class="co">](https://github.com/js229/Vennerable)</span>.</span>
<span id="cb18-165"><a href="#cb18-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-168"><a href="#cb18-168" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb18-169"><a href="#cb18-169" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> tot[<span class="dv">1</span>]</span>
<span id="cb18-170"><a href="#cb18-170" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> tot[<span class="dv">2</span>]</span>
<span id="cb18-171"><a href="#cb18-171" aria-hidden="true" tabindex="-1"></a>ab <span class="ot">&lt;-</span> tot[<span class="dv">3</span>]</span>
<span id="cb18-172"><a href="#cb18-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-173"><a href="#cb18-173" aria-hidden="true" tabindex="-1"></a>venn_obj <span class="ot">&lt;-</span> <span class="fu">createVennObj</span>(<span class="at">nSets =</span> <span class="dv">2</span>, <span class="at">sNames =</span> <span class="fu">c</span>(<span class="st">"m1043"</span>, <span class="st">"m1044"</span>), <span class="co"># names in order of first to last</span></span>
<span id="cb18-174"><a href="#cb18-174" aria-hidden="true" tabindex="-1"></a>                     <span class="at">sSizes =</span> <span class="fu">c</span>(<span class="dv">0</span>, a, b ,ab))</span>
<span id="cb18-175"><a href="#cb18-175" aria-hidden="true" tabindex="-1"></a>vp <span class="ot">&lt;-</span> <span class="fu">plotVenn</span>(<span class="at">nVennObj =</span> venn_obj)</span>
<span id="cb18-176"><a href="#cb18-176" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-177"><a href="#cb18-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-178"><a href="#cb18-178" aria-hidden="true" tabindex="-1"></a>This gets tedious with more than 2 sets. See code below. You should not make Venn Diagrams with more than 4 sets. An upset plot is better in that scenario.</span>
<span id="cb18-179"><a href="#cb18-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-182"><a href="#cb18-182" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb18-183"><a href="#cb18-183" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb18-184"><a href="#cb18-184" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-summary: "Show the code"</span></span>
<span id="cb18-185"><a href="#cb18-185" aria-hidden="true" tabindex="-1"></a><span class="co"># 3 sets -----</span></span>
<span id="cb18-186"><a href="#cb18-186" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> tot[<span class="dv">1</span>]</span>
<span id="cb18-187"><a href="#cb18-187" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> tot[<span class="dv">2</span>]</span>
<span id="cb18-188"><a href="#cb18-188" aria-hidden="true" tabindex="-1"></a>ab <span class="ot">&lt;-</span> tot[<span class="dv">3</span>]</span>
<span id="cb18-189"><a href="#cb18-189" aria-hidden="true" tabindex="-1"></a>c <span class="ot">&lt;-</span> tot[<span class="dv">4</span>]</span>
<span id="cb18-190"><a href="#cb18-190" aria-hidden="true" tabindex="-1"></a>ac <span class="ot">&lt;-</span> tot[<span class="dv">5</span>]</span>
<span id="cb18-191"><a href="#cb18-191" aria-hidden="true" tabindex="-1"></a>bc <span class="ot">&lt;-</span> tot[<span class="dv">6</span>]</span>
<span id="cb18-192"><a href="#cb18-192" aria-hidden="true" tabindex="-1"></a>abc <span class="ot">&lt;-</span> tot[<span class="dv">7</span>]</span>
<span id="cb18-193"><a href="#cb18-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-194"><a href="#cb18-194" aria-hidden="true" tabindex="-1"></a><span class="co"># 4 sets -------</span></span>
<span id="cb18-195"><a href="#cb18-195" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> tot[<span class="dv">1</span>]</span>
<span id="cb18-196"><a href="#cb18-196" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> tot[<span class="dv">2</span>]</span>
<span id="cb18-197"><a href="#cb18-197" aria-hidden="true" tabindex="-1"></a>ab <span class="ot">&lt;-</span> tot[<span class="dv">3</span>]Ve</span>
<span id="cb18-198"><a href="#cb18-198" aria-hidden="true" tabindex="-1"></a>c <span class="ot">&lt;-</span> tot[<span class="dv">4</span>]</span>
<span id="cb18-199"><a href="#cb18-199" aria-hidden="true" tabindex="-1"></a>ac <span class="ot">&lt;-</span> tot[<span class="dv">5</span>]</span>
<span id="cb18-200"><a href="#cb18-200" aria-hidden="true" tabindex="-1"></a>bc <span class="ot">&lt;-</span> tot[<span class="dv">6</span>]</span>
<span id="cb18-201"><a href="#cb18-201" aria-hidden="true" tabindex="-1"></a>abc <span class="ot">&lt;-</span> tot[<span class="dv">7</span>]</span>
<span id="cb18-202"><a href="#cb18-202" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> tot[<span class="dv">8</span>]</span>
<span id="cb18-203"><a href="#cb18-203" aria-hidden="true" tabindex="-1"></a>ad <span class="ot">&lt;-</span> tot[<span class="dv">9</span>]</span>
<span id="cb18-204"><a href="#cb18-204" aria-hidden="true" tabindex="-1"></a>bd <span class="ot">&lt;-</span> tot[<span class="dv">10</span>]</span>
<span id="cb18-205"><a href="#cb18-205" aria-hidden="true" tabindex="-1"></a>abd <span class="ot">&lt;-</span> tot[<span class="dv">11</span>]</span>
<span id="cb18-206"><a href="#cb18-206" aria-hidden="true" tabindex="-1"></a>cd <span class="ot">&lt;-</span> tot[<span class="dv">12</span>]</span>
<span id="cb18-207"><a href="#cb18-207" aria-hidden="true" tabindex="-1"></a>acd <span class="ot">&lt;-</span> tot[<span class="dv">13</span>]</span>
<span id="cb18-208"><a href="#cb18-208" aria-hidden="true" tabindex="-1"></a>bcd <span class="ot">&lt;-</span> tot[<span class="dv">14</span>]</span>
<span id="cb18-209"><a href="#cb18-209" aria-hidden="true" tabindex="-1"></a>abcd <span class="ot">&lt;-</span> tot[<span class="dv">15</span>]</span>
<span id="cb18-210"><a href="#cb18-210" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-211"><a href="#cb18-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-212"><a href="#cb18-212" aria-hidden="true" tabindex="-1"></a>This is a more automated way of reading in the data, here using 4 sets as an example.</span>
<span id="cb18-213"><a href="#cb18-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-216"><a href="#cb18-216" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb18-217"><a href="#cb18-217" aria-hidden="true" tabindex="-1"></a>venn <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"results/output_nepc.bedpe"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb18-218"><a href="#cb18-218" aria-hidden="true" tabindex="-1"></a>venn<span class="sc">$</span>alpha <span class="ot">&lt;-</span> <span class="fu">apply</span>(venn, <span class="dv">1</span>, <span class="cf">function</span>(x) {</span>
<span id="cb18-219"><a href="#cb18-219" aria-hidden="true" tabindex="-1"></a>  sets <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>)</span>
<span id="cb18-220"><a href="#cb18-220" aria-hidden="true" tabindex="-1"></a>  selected_sets <span class="ot">&lt;-</span> sets[<span class="fu">which</span>(x <span class="sc">==</span> <span class="st">"X"</span>)]</span>
<span id="cb18-221"><a href="#cb18-221" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste</span>(selected_sets, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb18-222"><a href="#cb18-222" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb18-223"><a href="#cb18-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-224"><a href="#cb18-224" aria-hidden="true" tabindex="-1"></a>order_vector <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"0"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"AB"</span>, <span class="st">"C"</span>, <span class="st">"AC"</span>, <span class="st">"BC"</span>, <span class="st">"ABC"</span>, <span class="st">"D"</span>, <span class="st">"AD"</span>, <span class="st">"BD"</span>, <span class="st">"ABD"</span>, <span class="st">"CD"</span>, <span class="st">"ACD"</span>, <span class="st">"BCD"</span>, <span class="st">"ABCD"</span>)</span>
<span id="cb18-225"><a href="#cb18-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-226"><a href="#cb18-226" aria-hidden="true" tabindex="-1"></a>venn<span class="sc">$</span>alpha <span class="ot">&lt;-</span> <span class="fu">factor</span>(venn<span class="sc">$</span>alpha, <span class="at">levels =</span> order_vector)</span>
<span id="cb18-227"><a href="#cb18-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-228"><a href="#cb18-228" aria-hidden="true" tabindex="-1"></a><span class="co"># Sort the data frame based on the factor levels</span></span>
<span id="cb18-229"><a href="#cb18-229" aria-hidden="true" tabindex="-1"></a>venn <span class="ot">&lt;-</span> venn[<span class="fu">order</span>(venn<span class="sc">$</span>alpha), ]</span>
<span id="cb18-230"><a href="#cb18-230" aria-hidden="true" tabindex="-1"></a>venn_plot <span class="ot">&lt;-</span> <span class="fu">Venn</span>(<span class="at">SetNames =</span> <span class="fu">c</span>(<span class="st">"93"</span>, <span class="st">"145.1"</span>, <span class="st">"145.2"</span>, <span class="st">"nci"</span>), </span>
<span id="cb18-231"><a href="#cb18-231" aria-hidden="true" tabindex="-1"></a>                  <span class="at">Weight =</span> <span class="fu">c</span>(<span class="dv">0</span>, venn<span class="sc">$</span>Features))</span>
<span id="cb18-232"><a href="#cb18-232" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(venn_plot, <span class="at">type =</span> <span class="st">"ellipses"</span>)</span>
<span id="cb18-233"><a href="#cb18-233" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-234"><a href="#cb18-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-235"><a href="#cb18-235" aria-hidden="true" tabindex="-1"></a>Creating upset plots <span class="ot">&lt;https://github.com/hms-dbmi/UpSetR&gt;</span></span>
<span id="cb18-236"><a href="#cb18-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-239"><a href="#cb18-239" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb18-240"><a href="#cb18-240" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(UpSetR)</span>
<span id="cb18-241"><a href="#cb18-241" aria-hidden="true" tabindex="-1"></a>venn<span class="sc">$</span>Sets <span class="ot">&lt;-</span> <span class="fu">apply</span>(venn[, <span class="sc">-</span><span class="dv">1</span>], <span class="dv">1</span>, <span class="cf">function</span>(x) {</span>
<span id="cb18-242"><a href="#cb18-242" aria-hidden="true" tabindex="-1"></a>  sets <span class="ot">&lt;-</span> <span class="fu">colnames</span>(venn)[<span class="dv">2</span><span class="sc">:</span><span class="dv">17</span>]</span>
<span id="cb18-243"><a href="#cb18-243" aria-hidden="true" tabindex="-1"></a>  selected_sets <span class="ot">&lt;-</span> sets[<span class="fu">which</span>(x <span class="sc">==</span> <span class="st">"X"</span>)]</span>
<span id="cb18-244"><a href="#cb18-244" aria-hidden="true" tabindex="-1"></a>  <span class="fu">paste</span>(selected_sets, <span class="at">collapse =</span> <span class="st">"&amp;"</span>)</span>
<span id="cb18-245"><a href="#cb18-245" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb18-246"><a href="#cb18-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-247"><a href="#cb18-247" aria-hidden="true" tabindex="-1"></a>upset_input <span class="ot">&lt;-</span> <span class="fu">c</span>(venn<span class="sc">$</span>Features)</span>
<span id="cb18-248"><a href="#cb18-248" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(upset_input) <span class="ot">&lt;-</span> venn<span class="sc">$</span>Sets</span>
<span id="cb18-249"><a href="#cb18-249" aria-hidden="true" tabindex="-1"></a><span class="fu">upset</span>(<span class="fu">fromExpression</span>(upset_input), <span class="at">nsets =</span> <span class="dv">16</span>, <span class="at">nintersects =</span> <span class="dv">100</span>, <span class="at">number.angles =</span> <span class="dv">45</span>)</span>
<span id="cb18-250"><a href="#cb18-250" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-251"><a href="#cb18-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-252"><a href="#cb18-252" aria-hidden="true" tabindex="-1"></a>A sample script to automate this process. I haven't incorporated the above part into this script.</span>
<span id="cb18-253"><a href="#cb18-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-254"><a href="#cb18-254" aria-hidden="true" tabindex="-1"></a><span class="in">```{bash filename="mergePeaksandplot.sh"}</span></span>
<span id="cb18-255"><a href="#cb18-255" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb18-256"><a href="#cb18-256" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load homer/4.10</span>
<span id="cb18-257"><a href="#cb18-257" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load R/4.0.3</span>
<span id="cb18-258"><a href="#cb18-258" aria-hidden="true" tabindex="-1"></a><span class="ex">mergePeaks</span> control_peaks.narrowPeak purturbed_peaks.narrowPeak <span class="at">-prefix</span> ov <span class="at">-venn</span> venn.txt</span>
<span id="cb18-259"><a href="#cb18-259" aria-hidden="true" tabindex="-1"></a><span class="co"># n_way script, venn.txt, last sample label, middle samples label, first sample label, txt file name (no extension)</span></span>
<span id="cb18-260"><a href="#cb18-260" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">--vanilla</span> venn2way.R venn.txt purturbed control venn</span>
<span id="cb18-261"><a href="#cb18-261" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-262"><a href="#cb18-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-263"><a href="#cb18-263" aria-hidden="true" tabindex="-1"></a>The accompanying R file:</span>
<span id="cb18-264"><a href="#cb18-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-265"><a href="#cb18-265" aria-hidden="true" tabindex="-1"></a><span class="in">```{r filename="venn2way.R"}</span></span>
<span id="cb18-266"><a href="#cb18-266" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb18-267"><a href="#cb18-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-268"><a href="#cb18-268" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Vennerable)</span>
<span id="cb18-269"><a href="#cb18-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-270"><a href="#cb18-270" aria-hidden="true" tabindex="-1"></a>args <span class="ot">&lt;-</span> <span class="fu">commandArgs</span>(<span class="at">trailingOnly =</span> <span class="cn">TRUE</span>)</span>
<span id="cb18-271"><a href="#cb18-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-272"><a href="#cb18-272" aria-hidden="true" tabindex="-1"></a>venn <span class="ot">&lt;-</span> <span class="fu">read.table</span>(args[<span class="dv">1</span>], <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>)</span>
<span id="cb18-273"><a href="#cb18-273" aria-hidden="true" tabindex="-1"></a>tot <span class="ot">&lt;-</span> venn[[Total]]</span>
<span id="cb18-274"><a href="#cb18-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-275"><a href="#cb18-275" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> tot[<span class="dv">1</span>]</span>
<span id="cb18-276"><a href="#cb18-276" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> tot[<span class="dv">2</span>]</span>
<span id="cb18-277"><a href="#cb18-277" aria-hidden="true" tabindex="-1"></a>ab <span class="ot">&lt;-</span> tot[<span class="dv">3</span>]</span>
<span id="cb18-278"><a href="#cb18-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-279"><a href="#cb18-279" aria-hidden="true" tabindex="-1"></a>venner <span class="ot">&lt;-</span> <span class="fu">Venn</span>(<span class="at">SetNames =</span> <span class="fu">c</span>(args[<span class="dv">3</span>], args[<span class="dv">2</span>]), <span class="co"># opposite labelling to nVennR, going from last sample to first</span></span>
<span id="cb18-280"><a href="#cb18-280" aria-hidden="true" tabindex="-1"></a>               <span class="at">Weight =</span> <span class="fu">c</span>(<span class="dv">0</span>, b, a, ab))</span>
<span id="cb18-281"><a href="#cb18-281" aria-hidden="true" tabindex="-1"></a><span class="fu">png</span>(<span class="fu">paste0</span>(args[<span class="dv">4</span>], <span class="st">".png"</span>))</span>
<span id="cb18-282"><a href="#cb18-282" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(venner)</span>
<span id="cb18-283"><a href="#cb18-283" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb18-284"><a href="#cb18-284" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-285"><a href="#cb18-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-286"><a href="#cb18-286" aria-hidden="true" tabindex="-1"></a>To make Venn Diagrams that are more accurately weighted, use <span class="co">[</span><span class="ot">nVennR</span><span class="co">](https://cran.r-project.org/web/packages/nVennR/index.html)</span>. Sadly, it looks like it's been removed from CRAN since the last time I've installed it.</span>
<span id="cb18-287"><a href="#cb18-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-288"><a href="#cb18-288" aria-hidden="true" tabindex="-1"></a>If you have replicates, overlapping peaks can be obtained via packages like <span class="co">[</span><span class="ot">MAnorm2</span><span class="co">](https://cran.r-project.org/web/packages/MAnorm2/index.html)</span>. Read more about dealing with replicates <span class="co">[</span><span class="ot">here</span><span class="co">](https://ro-che.info/articles/2018-07-11-chip-seq-consensus)</span>.</span>
<span id="cb18-289"><a href="#cb18-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-290"><a href="#cb18-290" aria-hidden="true" tabindex="-1"></a><span class="fu">### Motif Enrichment</span></span>
<span id="cb18-291"><a href="#cb18-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-292"><a href="#cb18-292" aria-hidden="true" tabindex="-1"></a>Homer <span class="in">`mergePeaks`</span> is able to take the narrowPeaks format as input, but if you're doing homer <span class="in">`mergePeaks`</span> first and then <span class="in">`findMotifs`</span>, you need to change the Homer <span class="in">`mergePeaks`</span> output to exclude the header line first. This also applies to deepTools heatmaps.</span>
<span id="cb18-293"><a href="#cb18-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-296"><a href="#cb18-296" aria-hidden="true" tabindex="-1"></a><span class="in">```{bash}</span></span>
<span id="cb18-297"><a href="#cb18-297" aria-hidden="true" tabindex="-1"></a><span class="ex">findMotifsGenome.pl</span> peak_or_bed /projects/p20023/Viriya/software/Homer4.10/data/genomes/hg38/ output_dir <span class="at">-size</span> 200 </span>
<span id="cb18-298"><a href="#cb18-298" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-299"><a href="#cb18-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-300"><a href="#cb18-300" aria-hidden="true" tabindex="-1"></a>-size 200 is the default, and is calculated from the center of the peaks. If your peaks are bigger and you wish to use the entire region, use <span class="in">`-size given`</span>. However, when the regions are too large, motifs will not be significantly enriched.</span>
<span id="cb18-301"><a href="#cb18-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-302"><a href="#cb18-302" aria-hidden="true" tabindex="-1"></a><span class="fu">### Plotting Heatmaps</span></span>
<span id="cb18-303"><a href="#cb18-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-304"><a href="#cb18-304" aria-hidden="true" tabindex="-1"></a>Here I'm showing a sample <span class="co">[</span><span class="ot">deepTools</span><span class="co">](https://deeptools.readthedocs.io/en/latest/)</span> script to plot heatmaps. First you need bigwig files. <span class="in">`bamCoverage`</span> offers normalization by scaling factor, Reads Per Kilobase per Million mapped reads (RPKM), counts per million (CPM), bins per million mapped reads (BPM) and 1x depth (reads per genome coverage, RPGC).</span>
<span id="cb18-305"><a href="#cb18-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-308"><a href="#cb18-308" aria-hidden="true" tabindex="-1"></a><span class="in">```{bash}</span></span>
<span id="cb18-309"><a href="#cb18-309" aria-hidden="true" tabindex="-1"></a><span class="ex">bamCoverage</span> <span class="at">--bam</span> m<span class="va">${i}</span>_sorted.bam <span class="at">--outFileName</span> m<span class="va">${i}</span>.bw <span class="at">--normalizeUsing</span> CPM <span class="dt">\</span></span>
<span id="cb18-310"><a href="#cb18-310" aria-hidden="true" tabindex="-1"></a>--extendReads 200 <span class="at">--numberOfProcessors</span> 6 <span class="at">--binSize</span> 20</span>
<span id="cb18-311"><a href="#cb18-311" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-312"><a href="#cb18-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-313"><a href="#cb18-313" aria-hidden="true" tabindex="-1"></a>I find having paths defined at the top of the script makes it less likely to have mistakes.</span>
<span id="cb18-314"><a href="#cb18-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-317"><a href="#cb18-317" aria-hidden="true" tabindex="-1"></a><span class="in">```{bash}</span></span>
<span id="cb18-318"><a href="#cb18-318" aria-hidden="true" tabindex="-1"></a><span class="va">day3</span><span class="op">=</span><span class="st">"m1015_1055_hg38/foxa2/CPM"</span></span>
<span id="cb18-319"><a href="#cb18-319" aria-hidden="true" tabindex="-1"></a><span class="va">nci</span><span class="op">=</span><span class="st">"m1015_1055_hg38/CPM"</span></span>
<span id="cb18-320"><a href="#cb18-320" aria-hidden="true" tabindex="-1"></a><span class="va">chip</span><span class="op">=</span><span class="st">"../Viriya/analysis/foxa2/chip-seq"</span></span>
<span id="cb18-321"><a href="#cb18-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-322"><a href="#cb18-322" aria-hidden="true" tabindex="-1"></a><span class="ex">computeMatrix</span> reference-point <span class="at">--referencePoint</span> center <span class="at">-S</span> <span class="dt">\</span></span>
<span id="cb18-323"><a href="#cb18-323" aria-hidden="true" tabindex="-1"></a><span class="va">$day3</span>/m674.bw <span class="va">$day3</span>/m676.bw <span class="va">$tf</span>/m914.bw <span class="va">$tf</span>/m921.bw <span class="va">$nci</span>/m1043.bw <span class="va">$nci</span>/m1044.bw <span class="dt">\</span></span>
<span id="cb18-324"><a href="#cb18-324" aria-hidden="true" tabindex="-1"></a>-R <span class="va">$chip</span>/clustering/output/1.bed <span class="dt">\</span></span>
<span id="cb18-325"><a href="#cb18-325" aria-hidden="true" tabindex="-1"></a><span class="va">$chip</span>/clustering/output/2.bed <span class="dt">\</span></span>
<span id="cb18-326"><a href="#cb18-326" aria-hidden="true" tabindex="-1"></a><span class="va">$chip</span>/clustering/output/3.bed <span class="dt">\</span></span>
<span id="cb18-327"><a href="#cb18-327" aria-hidden="true" tabindex="-1"></a><span class="va">$chip</span>/clustering/output/4.bed <span class="dt">\</span></span>
<span id="cb18-328"><a href="#cb18-328" aria-hidden="true" tabindex="-1"></a>-a 3000 <span class="at">-b</span> 3000 <span class="at">-o</span> <span class="va">$chip</span>/heatmaps/scaled_FOXA2_sites.npz <span class="dt">\</span></span>
<span id="cb18-329"><a href="#cb18-329" aria-hidden="true" tabindex="-1"></a>--samplesLabel m674 m676 m914 m921 m1043 m1044 <span class="dt">\</span></span>
<span id="cb18-330"><a href="#cb18-330" aria-hidden="true" tabindex="-1"></a>-p max <span class="at">--blackListFileName</span> <span class="va">$chip</span>/ENCFF356LFX.bed.gz</span>
<span id="cb18-331"><a href="#cb18-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-332"><a href="#cb18-332" aria-hidden="true" tabindex="-1"></a><span class="ex">plotHeatmap</span> <span class="at">-m</span> <span class="va">$chip</span>/FOXA2_6_samples/heatmaps/scaled_FOXA2_sites.npz <span class="dt">\</span></span>
<span id="cb18-333"><a href="#cb18-333" aria-hidden="true" tabindex="-1"></a>-o <span class="va">$chip</span>/FOXA2_6_samples/heatmaps/scaled_FOXA2_sites.pdf <span class="dt">\</span></span>
<span id="cb18-334"><a href="#cb18-334" aria-hidden="true" tabindex="-1"></a>--colorMap Blues</span>
<span id="cb18-335"><a href="#cb18-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-336"><a href="#cb18-336" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-337"><a href="#cb18-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-338"><a href="#cb18-338" aria-hidden="true" tabindex="-1"></a><span class="in">`--sortUsingSamples`</span> is not 0 indexed. 1st sample is 1. <span class="in">`--sortUsingSamples`</span> is also usable in plotHeatmap so we can save that to there for more flexible plotting.</span>
<span id="cb18-339"><a href="#cb18-339" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-340"><a href="#cb18-340" aria-hidden="true" tabindex="-1"></a><span class="in">`--regionsLabel`</span>: "xxx binding sites n=2390" put it in quotes if you don't want to key in escape too many times</span>
<span id="cb18-341"><a href="#cb18-341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-342"><a href="#cb18-342" aria-hidden="true" tabindex="-1"></a><span class="in">`--sampleLabels`</span>: if you're sure you're not going to change any labels, put it in makeheatmap step because fewer things to type when redoing the heatmap multiple times due to scaling etc, therefore fewer mistakes. otherwise put it in plotHeatmap.</span>
<span id="cb18-343"><a href="#cb18-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-344"><a href="#cb18-344" aria-hidden="true" tabindex="-1"></a>In legends: "Pol II" spaces and parantheses need to be escaped, but not + signs.</span>
<span id="cb18-345"><a href="#cb18-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-346"><a href="#cb18-346" aria-hidden="true" tabindex="-1"></a>Use BED6 format annd not the BED3 if you care about strandedness <span class="ot">&lt;https://github.com/deeptools/deepTools/issues/886&gt;</span>.</span>
<span id="cb18-347"><a href="#cb18-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-348"><a href="#cb18-348" aria-hidden="true" tabindex="-1"></a>To get sorted output bed files, use the <span class="in">`--sortedOutRegions`</span> at the plotHeatmap step, not computeMatrix.\</span>
<span id="cb18-349"><a href="#cb18-349" aria-hidden="true" tabindex="-1"></a>Use <span class="in">`--clusterUsingSamples`</span> for a more robust delineation.</span>
<span id="cb18-350"><a href="#cb18-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-351"><a href="#cb18-351" aria-hidden="true" tabindex="-1"></a><span class="fu">### Genomic Annotation</span></span>
<span id="cb18-352"><a href="#cb18-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-353"><a href="#cb18-353" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">ChIPSeeker</span><span class="co">](https://bioconductor.org/packages/release/bioc/html/ChIPseeker.html)</span></span>
<span id="cb18-354"><a href="#cb18-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-357"><a href="#cb18-357" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb18-358"><a href="#cb18-358" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb18-359"><a href="#cb18-359" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TxDb.Hsapiens.UCSC.hg38.knownGene)</span>
<span id="cb18-360"><a href="#cb18-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-361"><a href="#cb18-361" aria-hidden="true" tabindex="-1"></a>promoter <span class="ot">&lt;-</span> <span class="fu">getPromoters</span>(<span class="at">TxDb=</span>txdb, <span class="at">upstream=</span><span class="dv">3000</span>, <span class="at">downstream=</span><span class="dv">3000</span>)</span>
<span id="cb18-362"><a href="#cb18-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-363"><a href="#cb18-363" aria-hidden="true" tabindex="-1"></a>mxxx <span class="ot">&lt;-</span> <span class="fu">readPeakFile</span>(<span class="st">"/path_to_narrowPeak/mxxx_peaks.narrowPeak"</span>)</span>
<span id="cb18-364"><a href="#cb18-364" aria-hidden="true" tabindex="-1"></a>tagMatrix <span class="ot">&lt;-</span> <span class="fu">getTagMatrix</span>(mxxx, <span class="at">windows =</span> promoter) <span class="co"># usually needs an interactive job</span></span>
<span id="cb18-365"><a href="#cb18-365" aria-hidden="true" tabindex="-1"></a>peakAnno <span class="ot">&lt;-</span> <span class="fu">annotatePeak</span>(mxxx, <span class="at">tssRegion =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">3000</span>,<span class="dv">3000</span>), <span class="at">TxDb =</span> txdb, <span class="at">annoDb =</span> <span class="st">"org.Hs.eg.db"</span>)</span>
<span id="cb18-366"><a href="#cb18-366" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAnnoPie</span>(peakAnno, <span class="at">main =</span> <span class="st">"mxxx"</span>, <span class="at">line =</span> <span class="sc">-</span><span class="dv">6</span>)</span>
<span id="cb18-367"><a href="#cb18-367" aria-hidden="true" tabindex="-1"></a><span class="fu">vennpie</span>(peakAnno, <span class="at">r =</span> <span class="fl">0.1</span>)</span>
<span id="cb18-368"><a href="#cb18-368" aria-hidden="true" tabindex="-1"></a><span class="fu">upsetplot</span>(peakAnno) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">"mxxx"</span>)</span>
<span id="cb18-369"><a href="#cb18-369" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAnnoBar</span>(peakAnno)</span>
<span id="cb18-370"><a href="#cb18-370" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb18-371"><a href="#cb18-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-372"><a href="#cb18-372" aria-hidden="true" tabindex="-1"></a><span class="in">`annotatePeak`</span> will default to chosing one gene per region. This is fine for small binding sites but for larger regions use <span class="in">`seq2gene`</span> as it will map genomic regions in a many-to-many manner.</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>